pages/api/submit.ts

```diff
@@ -19,44 +38,86 @@ async function serverSubmit(req: NextApiRequest, res: NextApiResponse) {
+  // updated and we can proceed to the next page.
+  let page = currentPage;
   let queryValues = "";
+  let addresses;
+  let results;
+  let errors;
 
-  // TODO: run values through validation. After the new form field values
-  // are validated, then merge all together:
-  newSubmittedValues = { ...newSubmittedValues, ...existingValues };
-  for (const [key, value] of Object.entries(newSubmittedValues)) {
-    queryValues += `&${key}=${value}`;
-  }
-  // A switch statement so later on we can do form value validation on specific
-  // fields based on the current page.
   switch (currentPage) {
     case "personal":
-      page = "location";
+      // Validate the form values for this specific step.
+      errors = validatePersonalFormData({}, newSubmittedValues);
+      if (isEmpty(errors)) {
+        page = "location";
+      }
       break;
     case "location":
-      page = "workAddress";
+      addresses = constructAddresses(newSubmittedValues);
+      errors = validateAddressFormData({}, addresses);
+      if (isEmpty(errors)) {
+        if (newSubmittedValues.location !== "nyc") {
+          page = "workAddress";
+        } else {
+          page = "account";
+        }
+      }
       break;
     case "workAddress":
-      page = "address-verification";
-      // TODO: Do address API call here.
-      break;
-    case "addressVerification":
-      page = "account";
+      // We need to add the existing home address values since
+      // `validateAddressFormData` checks for both home and work addresses.
+      addresses = constructAddresses({
+        ...newSubmittedValues,
+        ...existingValues,
+      });
+      errors = validateAddressFormData({}, addresses);
+      if (isEmpty(errors)) {
+        page = "account";
+      }
       break;
     case "account":
-      page = "review";
+      errors = validateAccountFormData({}, newSubmittedValues);
+      if (isEmpty(errors)) {
+        page = "review";
+      }
       break;
     case "review":
-      page = "congrats";
-      // TODO: Call the `createPatron` function with all the form values here.
+      try {
+        // All the data values are in the `existingValues` variable since
+        // no new data is submitted on the "review" page. When JS is not on,
+        // checkbox values are submitted as "on" and "off". We need to convert
+        // the values into booleans for the API to understand.
+        for (const [key, value] of Object.entries(existingValues)) {
+          if (value === "on") {
+            existingValues[key] = true;
+          } else if (value === "off") {
+            existingValues[key] = false;
+          }
+        }
+        // If the API call was successful and a patron account was created,
+        // go to the congrats page and display the results.
+        results = await callPatronAPI(existingValues);
+        page = "congrats";
+      } catch (error) {
+        errors = error;
+      }
```

Comment:
This is the most complicated part since here we make the actual API call to create an ILS account through `callPatronAPI`.